### PR TITLE
New commands: expressive-create-middleware and metacommand "expressive"

### PR DESCRIPTION
**Pull request by weierophinney**
This adds a new command, `expressive-create-middleware`, which allows you to create a PSR-15 middleware classfile for the provided fully-qualified class name:

```bash
$ ./vendor/bin/expressive-create-middleware "Foo\BarMiddleware"
```

It will create it in a directory based on matching the namespace of the requested class against defined PSR-4 autoloaders for the current project.

This patch also creates the metacommand `expressive`, which allows you to invoke all other commands in this project:

```bash
$ ./vendor/bin/expressive module create Foo
$ ./vendor/bin/expressive create-middleware "Foo\BarMiddleware"
```

When invoked with no arguments or a `help` argument, it lists available commands. It also allows you to access help for individual commands using the syntax `expressive help <command>`.

### CODE REVIEW

File: src/CreateMiddleware/Command.php [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    private $helpArgs = ['--help', '-h', 'help'];
+
+    /**
+     * @var ConsoleHelper
+     */
+    private $console;
+
+    /**
+     * @param string $command Script that is invoking the command.
+     * @param ConsoleHelper $console
```
Comment (michalbundyra) [inline src/CreateMiddleware/Command.php]: Missing `null` as allowed value type.

File: src/CreateMiddleware/CreateMiddleware.php [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        // $response = $delegate->process($request);
+    }
+}
+EOS;
+
+    /**
+     * @param string $class
+     * @param string $projectRoot
+     * @return string
+     * @throws CreateMiddlewareException
```
Comment (michalbundyra) [inline src/CreateMiddleware/CreateMiddleware.php:45]: I can't see that exception in this method. Should it be here?

File: src/CreateMiddleware/CreateMiddleware.php [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    private function getComposerAutoloaders($projectRoot)
+    {
+        $composerPath = sprintf('%s/composer.json', $projectRoot);
+        if (! file_exists($composerPath)) {
+            throw CreateMiddlewareException::missingComposerJson();
+        }
+
+        $composer = json_decode(file_get_contents($composerPath), true);
+
+        if (json_last_error() !== \JSON_ERROR_NONE) {
```
Comment (michalbundyra) [inline src/CreateMiddleware/CreateMiddleware.php:113]: Do we need `\` before constant name?

File: test/ExpressiveCommandTest.php [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        );
+    }
+
+    public function helpArguments()
+    {
+        return [
+            'no-args'               => [[]],
+            'long-arg'              => [['--help']],
+            'short-arg'             => [['-h']],
+            'command'               => [['help']],
```
Comment (michalbundyra) [inline test/ExpressiveCommandTest.php]: Can we have less spaces above before `=>`?

File: test/ExpressiveCommandTest.php [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        ];
+    }
+
+    /**
+     * @dataProvider helpArguments
```
Comment (michalbundyra) [inline test/ExpressiveCommandTest.php:42]: Please add PHPDoc for param `$args`.

File: test/ExpressiveCommandTest.php [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+            'post-migrate-original-messages' => [['migrate-original-messages', 'help'], 'migrate-original-messages'],
+            'post-module'                    => [['module', 'help'], 'expressive module'],
+            'post-pipeline-from-config'      => [['pipeline-from-config', 'help'], 'expressive pipeline-from-config'],
+            'post-scan-for-error-middleware' => [['scan-for-error-middleware', 'help'], 'expressive scan-for-error-middleware'],
+        ];
+        // @codingStandardsIgnoreEnd
+    }
+
+    /**
+     * @dataProvider commandHelpArguments
```
Comment (michalbundyra) [inline test/ExpressiveCommandTest.php:97]: As above - please add PHPDoc for params.

File: src/Module/Command.php [hunk 7]
```diff
@@ -18,7 +18,7 @@ class Command
     /**
      * @var string
      */
-    private $projectDir = '.';
+    public $projectDir = '.';
```
Comment (michalbundyra) [inline src/Module/Command.php:21]: Why this change?

File: src/CreateMiddleware/CreateMiddleware.php (hunk 2 above)
Comment (weierophinney) [reply src/CreateMiddleware/CreateMiddleware.php:45]: Yes. This is the only public method on an instance, so we need to report any exceptions that could be thrown by invoking it. In this particular case, it delegates to private methods that can throw this exception.

File: src/CreateMiddleware/CreateMiddleware.php (hunk 3 above)
Comment (weierophinney) [reply src/CreateMiddleware/CreateMiddleware.php:113]: It's an optimization:

- it ensures we're grabbing the global constant, and not one defined in this namespace.
- it prevents PHP from failing to resolve in the current namespace prior to looking for the constant in the global namespace.

File: src/Module/Command.php (hunk 7 above)
Comment (weierophinney) [reply src/Module/Command.php:21]: To make testing possible. See [this line in ExpressiveCommand](https://github.com/zendframework/zend-expressive-tooling/pull/22/files#diff-b240d71305c9faddf69a01e76b75fdabR158); without that, we cannot do test cases to ensure delegation works correctly for commands that utilize the filesystem.
